feat: use direct wire format -> LocalDate conversion without resorting to java.util.Date, java.util.Calendar, and default timezones

uschindler opened the pull request:
This fixes #2221:

I added a new method to TimestampUtils that directly converts the epoch days to a LocalDate. Interestingly this allows to not use the cryptic proleptic seconds hack. You can pass the plain epoch days from postgres, just added a few number of days -- actually the strange constant at beginning to `toJavaSecs()` method:

```
    // postgres epoc to java epoc
    secs += 946684800L;
```

divided by 86400 (to get it in days). The following code in toJavaSecs is not needed and returns broken results, because `LocalDate` also does its own Julian/Gregorian transformation.

I also beefed the test for LocalDate to cycle through all timezones (should not be needed, as timezone is not involved). I also added dates like 0001-01-01 to tests for LocalDate and LocalDateTime.

uschindler commented on pgjdbc/src/main/java/org/postgresql/jdbc/TimestampUtils.java:
```diff
@@ -1258,6 +1258,24 @@ public LocalDateTime toLocalDateTimeBin(byte[] bytes) throws PSQLException {
+	  }
+	  int days = ByteConverter.int4(bytes, 0);
+	  days += 10957L; // adapt the different Postgres Epoch
+	  return LocalDate.ofEpochDay(days);
```
Here I am not fully sure if we need the min/max checks. But Postgres Dates should always fit into LocalDate, so the check is IMHO obsolete. What do others think?

vlsi replied:
Check `org.postgresql.test.jdbc42.GetObject310InfinityTests` and https://www.postgresql.org/docs/9.0/datatype-datetime.html#AEN5647

I wonder how would you fit `-infinity` and `infinity` into `LocalDate` :)

vlsi replied:
Ok, PostgreSQL defines `-infinity` and `infinity` as MIN_INT32 and MAX_INT32: https://github.com/postgres/postgres/blob/27b77ecf9f4d5be211900eda54d8155ada50d696/src/include/utils/date.h#L34-L43
So wire format is exactly that: "min_int" vs "max_int"

vlsi replied:
Apparently, `days += 10957L` would overflow `int32`.

So I suggest treating `MAX_INT` and `MIN_INT` as a special case, and use `long` variable for `days`.
It is really sad that Java allows integer overflow in `days += 10957L`

uschindler replied:
Will add a test and implement this check.

vlsi replied:
The test already exists, no need to add it.
However, it would probably be interesting to test cases like "infinity - 1 day" (e.g. max_int32 - 1day).

uschindler replied:
OK, I fixed the overflow. Actually LocalDate.fromEpochDays takes a long, I just declared the variable in a wrong way (copynpaste - shitnwaste),
I also added the Integer.MIN_VALUE and Integer.MAX_VALUE checks in the same way like for LocalDateTime.

uschindler replied:
The `GetObject310InfinityTests` now also passes.

bokken commented:
I wonder if we should address /all/ of the date and time data types (not just LocalDate), as the discussion (and proposed change) seems pretty similar to where discussion on https://github.com/pgjdbc/pgjdbc/pull/1391 is going.

`the strange constant at beginning to toJavaSecs()`

This is the difference between the java/unix epoch of Jan 1, 1970 and the postgres epoch of Jan 1, 2000.

Another way to do this (which might be less cryptic) is to define a PG_EPOCH constant and do plus/minus operations from there.

`private static final LocalDateTime PG_EPOCH = LocalDateTime.of(2000, 1, 1, 0, 0);`

vlsi commented:
> I wonder if we should address /all/ of the date and time data types (not just LocalDate),

I wonder if going with `java.time.` causes performance degradation or not.

uschindler commented:
> > I wonder if we should address /all/ of the date and time data types (not just LocalDate),
> 
> I wonder if going with `java.time.` causes performance degradation or not.

At least not for this fix. The previous code with converting between all those old date/time and the reflection to get current timezone not allowed in Java 9+ for nonsense should make at least those faster. And this old-style cleandar magic is also not the fastest. The current code does:
- convert the Date to a new java.sql.Date with reflective access to get timezone, coping into calendar and modifying to local timezone
- convert the date back with the same timezone when appliyng toLocalDate

All not good and errors and bugs waiting everywhere!

bokken commented:
> >    I wonder if we should address /all/ of the date and time data types (not just LocalDate),
> 
> I wonder if going with java.time. causes performance degradation or not.

Based on a cursory glance at the implementation, I would suspect it is actually faster.

vlsi commented:
> At least not for this fix

That is no question. If user asks for `java.time.`, it is just faster to create one.

However, what @bokken  mentioned was "should we **always** build the value as `java.time.`, and then convert to `java.util` if user calls regular `ResultSet#getTimestamp` API. That change might have performance implications (both positive and negative).

uschindler commented:
I have not tested all stuff yet, so this issue is still WIP. Thanks for the comments, will update later.

I forgot to mark as Draft.

vlsi commented on pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java:
```diff
@@ -734,6 +734,30 @@ public int getConcurrency() throws SQLException {
+      }
+    }
+
+    return getTimestampUtils().toLocalDateTime(castNonNull(getString(i))).toLocalDate();
```
I wonder if we want to gradually remove `getTimestampUtils()`.
For instance, `byte[] -> LocalDate` conversion does not really need `TimestampUtils` instance. Should the method be static instead?

Then it would ease understanding of what `toLocalDateTime` could use, and it would avoid accidental use of shared state. WDYT?

uschindler replied:
I was thinking about that, too. Theoretically the TimestampUtils should only contain the legacy stuff. The java.time stuff could completely be moved to PgResultSet.
As first step I'd make the methods static, as suggested.

uschindler replied:
I would do this in a separate cleanup PR. I made it consistent with the LocalDateTime code.

uschindler commented:
I think we can't use java.time Apis for the legacy stuff, because we need to support custom calendars passed in to get date and others. Unless we can integrate that, we have to stay with the legacy code.

But we should make sure that if java.time is requested, we should directly deserialize to that. For speed and to keep time zones put of the game.

uschindler commented:
I changed the integer overflow and added infinity handling. The infinity tests also pass.

@bokken: I also added the "strange constants" precalculated using `java.time.Duration` as static final longs.

uschindler commented:
I looked into the other remianing legacy conversion (`TIMETZ -> OffsetDateTime`). This still uses legacy `java.sql.Time` at two places:

- https://github.com/pgjdbc/pgjdbc/blob/54e5598ab34ed791d7682cfd8700be3e8b7b37e3/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java#L687-L693
- https://github.com/pgjdbc/pgjdbc/blob/54e5598ab34ed791d7682cfd8700be3e8b7b37e3/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java#L705-L711

Both are not tested! I set a breakpoint in both if clauses and it was never hit!

The 2nd one looks obsolete, because the standard string Timestamp parser can also parse a `TIME` or `TIMETZ` only (the date will then be returned as 1970-01-01). So Lines 705 to 711 can be removed, the code below should be able to read it. But before doing that a test should be added. Especially for the special case of "24:00:00".

The first one requires a binary parser from `TIMETZ` to `java.time.OffsetTime`. This is easy to implement and I would like to also do this. But then I noticed that we have no support for doing `getObject(..., OffsetTime.class)` at all. What do others think, would this be a good addition?

I know, use of `TIMETZ` is discouraged by Postgres, but for full JDBC support of modern time APIs, `OffsetTime` should be supported as it also offsers a way to get the time including its timezone (as stored in database).

I could do those changes in this PR, but as alternative I would start working on it as a new PR, if this one is merged. What do others think? @bokken @vlsi 

Last sentence for @bokken: At a later stage we may not fully switch to java.time totally (see my comment above), but at least the parsing of string timestamps coming from database should possibly be implemented using java.time and the value-based helper classes `ParsedTimestamp` and so on should be removed in favor of `OffsetDateTime`. This may be another cleanup.

davecramer commented:
> I looked into the other remianing legacy conversion (`TIMETZ -> OffsetDateTime`). This still uses legacy `java.sql.Time` at two places:
> 
> * https://github.com/pgjdbc/pgjdbc/blob/54e5598ab34ed791d7682cfd8700be3e8b7b37e3/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java#L687-L693
> * https://github.com/pgjdbc/pgjdbc/blob/54e5598ab34ed791d7682cfd8700be3e8b7b37e3/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java#L705-L711
> 
> Both are not tested! I set a breakpoint in both if clauses and it was never hit!
> 
> The 2nd one looks obsolete, because the standard string Timestamp parser can also parse a `TIME` or `TIMETZ` only (the date will then be returned as 1970-01-01). So Lines 705 to 711 can be removed, the code below should be able to read it. But before doing that a test should be added. Especially for the special case of "24:00:00".
> 
> The first one requires a binary parser from `TIMETZ` to `java.time.OffsetTime`. This is easy to implement and I would like to also do this. But then I noticed that we have no support for doing `getObject(..., OffsetTime.class)` at all. What do others think, would this be a good addition?
> 
> I know, use of `TIMETZ` is discouraged by Postgres, but for full JDBC support of modern time APIs, `OffsetTime` should be supported as it also offsers a way to get the time including its timezone (as stored in database).
> 
> I could do those changes in this PR, but as alternative I would start working on it as a new PR, if this one is merged. What do others think? @bokken @vlsi
> 
> Last sentence for @bokken: At a later stage we may not fully switch to java.time totally (see my comment above), but at least the parsing of string timestamps coming from database should possibly be implemented using java.time and the value-based helper classes `ParsedTimestamp` and so on should be removed in favor of `OffsetDateTime`. This may be another cleanup.

I would prefer another PR. 
- moves this one along faster
- separate discussion, etc ....

davecramer commented on pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java:
```diff
@@ -734,6 +733,30 @@ public int getConcurrency() throws SQLException {
+        return getTimestampUtils().toLocalDateBin(value);
+      } else if (oid == Oid.TIMESTAMP) {
+        return getTimestampUtils().toLocalDateTimeBin(value).toLocalDate();
+      } else {
```
So why would we not want to return a localdate from a timestamp with time zone ?

uschindler replied:
The original code line 3699 did not supported this, too.

But I fully aggree, we can add it, is a trivial addition similar to this one. If we support it here, we must add it in line 372, too.

uschindler replied:
In `getLocalDateTime()` above it also does not support it.

davecramer replied:
Well this is a question for debate. The fact that it wasn't supported is somewhat orthogonal

uschindler replied:
At other places it says "xy is required by JDBC spec", maybe it did just implement the minimal set.

In my opinion the main getObject() method can easily support all types with and without offset. The problem with the TZ ones is that it is unclear to me if it should just strip time zone information or use default timezone during the conversion. But I haven't read the JDBC spec how this should work. So better throw exception instead of returning wrong values.

uschindler replied:
As said in my previous post the whole `getObject()` is also missing `OffsetTime` from `TIMETZ`. This would be a really nice addition.

uschindler commented:
> I think my proposal is that we pick a preferred internal "native" representation (a java.time type) for each of the pg types and then map how to do the conversions.

OK, this is a good approach. At least it makes it easy to correctly impement the mapping from wire byte arrays. It also does not need the julian/gregorian conversions, because the java.time API does it correctly (as shown by the LocalDateTime and LocalDate conversions).

I think we still need some massaging for the legacy types, where we may need to apply local timezone for the "Local", non-timezoned types. This is how they were originally defined and what also mekes them so horrible to use. But the timezone for this conversion can be picked from the passed Calendar instance to `getTimestamp()` & Co.

vlsi commented on pgjdbc/src/main/java/org/postgresql/jdbc/TimestampUtils.java:
```diff
@@ -63,6 +63,8 @@
   private static final LocalDate MIN_LOCAL_DATE = LocalDate.of(4713, 1, 1).with(ChronoField.ERA, IsoEra.BCE.getValue());
   private static final LocalDateTime MIN_LOCAL_DATETIME = MIN_LOCAL_DATE.atStartOfDay();
   private static final OffsetDateTime MIN_OFFSET_DATETIME = MIN_LOCAL_DATETIME.atOffset(ZoneOffset.UTC);
+  private static final Duration PG_EPOCH_DIFF = Duration.between(Instant.EPOCH, LocalDate.of(2000, 1, 1)
+      .atStartOfDay().toInstant(ZoneOffset.UTC));
```
```suggestion
  private static final Duration PG_EPOCH_DIFF =
      Duration.between(Instant.EPOCH, LocalDate.of(2000, 1, 1).atStartOfDay().toInstant(ZoneOffset.UTC));
```

No idea which one is better, however, having `.atStartOfDay` separated from `LocalDate.of` looks strange

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc42/GetObject310Test.java:
```diff
@@ -90,17 +90,72 @@ public void tearDown() throws SQLException {
-      assertEquals(localDate, rs.getObject(1, LocalDate.class));
+      stmt.executeUpdate(TestUtil.insertSQL("table1","date_column","DATE '" + date + "'"));
+
+      ResultSet rs = stmt.executeQuery(TestUtil.selectSQL("table1", "date_column"));
+      try {
```
```suggestion
      try (ResultSet rs = stmt.executeQuery(TestUtil.selectSQL("table1", "date_column")); ) {
```

uschindler replied:
The try with resources should be used in whole test. Maybe as a cleanup issue later.

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc42/GetObject310Test.java:
```diff
@@ -90,17 +90,72 @@ public void tearDown() throws SQLException {
+
+  public void localDate(ZoneId zoneId, String date) throws SQLException {
+    TimeZone.setDefault(TimeZone.getTimeZone(zoneId));
+    Statement stmt = con.createStatement();
     try {
```
```suggestion
    try (Statement stmt = con.createStatement(); ) {
```

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc42/GetObject310Test.java:
```diff
@@ -90,17 +90,72 @@ public void tearDown() throws SQLException {
+        LocalDate localDate = LocalDate.parse(date);
+        assertEquals(localDate, rs.getObject("date_column", LocalDate.class));
+        assertEquals(localDate, rs.getObject(1, LocalDate.class));
+      } finally {
+        rs.close();
+      }
```
```suggestion
      }
```

vlsi reviewed (approved):
Looks good to me modulo small style issues. Thanks for the PR.

vlsi commented:
Will commit the changes shortly unless someone does that earlier

vlsi commented on pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java:
```diff
@@ -734,6 +733,30 @@ public int getConcurrency() throws SQLException {
+      } else {
+        throw new PSQLException(
+            GT.tr("Cannot convert the column of type {0} to requested type {1}.",
+                Oid.toString(oid), "date"),
```
```suggestion
                Oid.toString(oid), "java.time.LocalDate"),
```

uschindler commented:
Thanks for review. I applied your suggestions.

uschindler commented:
Thanks for merging. I will proceed with a separate PR about TIME and TIMETZ which gas same problem like described here. This will also add support for OffsetTime then which is completely missing.

vlsi commented:
Just in case, I see you follow the code style that was there in the classes before (I really appreciate that!), however, please, don't hesitate to use Java 8+ features like `try-with-resources` in the newly created code.

Historically the driver was Java 6+, so the code idioms are still from the old days. I think the newly added code could just use 8+, especially for the cases like `try-with-resources`, lambdas in assertions.

uschindler commented:
> however, please, don't hesitate to use Java 8+ features like try-with-resources in the newly created code.

The background here was: I copypasted the `LocalDateTime`-iterate-over-timezones test and modified it - I agree, I could have changed it! :-)
